federation/pkg/federation-controller/dns/dns.go

```diff
@@ -14,90 +14,223 @@ See the License for the specific language governing permissions and
+		glog.Infof("DNS zone %q successfully created.  Note that DNS resolution will not work until you have "+
+			"registered this name with a DNS registrar and they have changed the authoritative name "+
+			"servers for your domain to point to your DNS provider.", zone.Name())
+	}
+	if len(matchingZones) > 1 {
+		return fmt.Errorf("Multiple matching DNS zones found for %q; please specify zoneID", dc.zoneName)
+	}
+	dc.dnsZone, err = getDNSZone(dc.zoneName, dc.zoneID, dc.dnsZones)
+	if err != nil {
+		return fmt.Errorf("Get DNS Zone failed: %v", err)
 	}
-	return zoneEndpoints, regionEndpoints, globalEndpoints, nil
+	return nil
 }
 
-// getClusterZoneNames returns the name of the zones (and the region) where the specified cluster exists (e.g. zones "us-east1-c" on GCE, or "us-east-1b" on AWS)
-func (s *ServiceController) getClusterZoneNames(clusterName string) (zones []string, region string, err error) {
-	client, ok := s.clusterCache.clientMap[clusterName]
-	if !ok {
-		return nil, "", fmt.Errorf("Cluster cache does not contain entry for cluster %s", clusterName)
+func (dc *DNSController) Run(workers int, stopChan <-chan struct{}) {
+	for i := 0; i < workers; i++ {
+		go wait.Until(dc.worker, time.Second, stopChan)
 	}
-	if client.cluster == nil {
-		return nil, "", fmt.Errorf("Cluster cache entry for cluster %s is nil", clusterName)
+	go dc.serviceInformerController.Run(stopChan)
+}
+
+func (sc *DNSController) worker() {
+	for {
+		item, quit := sc.workQueue.Get()
+		if quit {
+			return
+		}
+		service := item.(*apiv1.Service)
+		sc.reconcileDNS(service)
+		sc.workQueue.Done(item)
 	}
-	return client.cluster.Status.Zones, client.cluster.Status.Region, nil
 }
 
-// getServiceDnsSuffix returns the DNS suffix to use when creating federated-service DNS records
-func (s *ServiceController) getServiceDnsSuffix() (string, error) {
-	return s.serviceDnsSuffix, nil
+// getUplevelCName returns one level higher CNAME for a dnsName
+// if dnsName is of zone level, then returns region CNAME
+// if dnsName is of region level, then returns global CNAME
+// if dnsName is of global level, then returns "" empty string
+func (dc *DNSController) getUplevelCName(dnsName, prefix string) (uplevelCname string) {
+	prefixSlice := strings.Split(prefix, ".")
+	dnsNameSlice := strings.Split(dnsName, ".")
+	itemToDelete := len(prefixSlice)
+	dnsNameSlice = append(dnsNameSlice[:itemToDelete], dnsNameSlice[itemToDelete+1:]...)
+
+	uplevelCname = strings.Join(dnsNameSlice, ".")
+	if !strings.HasSuffix(uplevelCname, dc.serviceDNSSuffix) {
+		uplevelCname = ""
+	}
+	return uplevelCname
 }
 
-// getDnsZones returns the DNS zones matching dnsZoneName and dnsZoneID (if specified)
-func getDnsZones(dnsZoneName string, dnsZoneID string, dnsZonesInterface dnsprovider.Zones) ([]dnsprovider.Zone, error) {
+// getDNSZones returns the DNS zones matching dnsZoneName and dnsZoneID (if specified)
+func getDNSZones(dnsZoneName string, dnsZoneID string, dnsZonesInterface dnsprovider.Zones) (
```

Comment:
agreed, i did not put enough focus on how the diff would look like as the diff looked inconsistent because of lot of change, but wherever possible should have kept the original code instead enforcing line wrapping strictly. will take care as much as possible.